pyflakes/checker.py

```diff
@@ -1074,17 +1085,33 @@ def NAME(self, node):
         """
         Handle occurrence of Name (which can be a load/store/delete access.)
         """
+        global_scope_index = 1 if self._in_doctest() else 0
+        global_scope = self.scopeStack[global_scope_index]
         # Locate the name in locals / function / globals scopes.
         if isinstance(node.ctx, (ast.Load, ast.AugLoad)):
             self.handleNodeLoad(node)
             if (node.id == 'locals' and isinstance(self.scope, FunctionScope)
                     and isinstance(node.parent, ast.Call)):
                 # we are doing locals() call in current scope
                 self.scope.usesLocals = True
+            if (isinstance(self.scope, FunctionScope) and
+                    node.id in self.scope.global_names):
+                if node.id not in global_scope:
+                    self.report(messages.UndefinedName, node, node.id)
         elif isinstance(node.ctx, (ast.Store, ast.AugStore)):
             self.handleNodeStore(node)
+            if (isinstance(self.scope, FunctionScope) and
```

Comment:
this should go in `handleNodeStore`, where there is already logic regarding globals